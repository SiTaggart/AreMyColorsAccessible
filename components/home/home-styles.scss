@import '../../styles/utils';

.home {
  position: relative;

  @include breakpoint($small-screen) {
    min-height: 85%;
  }
}
