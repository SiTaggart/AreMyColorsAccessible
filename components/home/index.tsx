import React, { Component, ReactElement } from 'react';
import { SiteData, ColorCombinationTypes } from '../../types';
import Container from '../layouts/container';
import Layout from '../layouts/layout';
import Results from '../results';
import ColorInputs from '../colorInputs';

export interface HomeProps {
  handleBackgroundColorInputChange: (value: string) => void;
  handleBackgroundColorSliderChange: (hex: string) => void;
  handleTextColorInputChange: (value: string) => void;
  handleTextColorSliderChange: (hex: string) => void;
  siteData: SiteData;
}

class Home extends Component<HomeProps, {}> {
  public render(): ReactElement {
    const siteData: SiteData = this.props.siteData;
    const colorInfo: Partial<ColorCombinationTypes> = siteData.colorCombos[0].combinations[0];

    return (
      <Container variant="home">
        <Layout variant="small">
          <Results
            accessibility={colorInfo.accessibility}
            contrast={colorInfo.contrast}
            isLight={siteData.isLight}
          />
        </Layout>
        <Layout variant="large">
          <ColorInputs
            background={siteData.background}
            colorCombos={siteData.colorCombos}
            handleBackgroundColorInputChange={this.props.handleBackgroundColorInputChange}
            handleBackgroundColorSliderChange={this.props.handleBackgroundColorSliderChange}
            handleTextColorInputChange={this.props.handleTextColorInputChange}
            handleTextColorSliderChange={this.props.handleTextColorSliderChange}
            isLight={siteData.isLight}
            textColor={siteData.textColor}
          />
        </Layout>
      </Container>
    );
  }
}

export default Home;
