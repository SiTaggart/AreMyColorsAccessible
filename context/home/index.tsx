import * as React from 'react';
import Color from 'color';
import isEmpty from 'lodash/isEmpty';
import debounce from 'lodash/debounce';
import qs from 'query-string';
import ColorCombos, { ColorCombosTypes, ColorCombinationTypes } from 'color-combos';
import { SiteData } from '../../types';

export interface HomeContextInterface {
  siteData: SiteData;
  handleBackgroundColorInputChange: (value: string) => void;
  handleTextColorInputChange: (value: string) => void;
}

interface SiteDataProviderProps {
  children?: React.ReactElement;
  initialSiteData: SiteData | {};
}

const setInitialContext = (initialSiteData: SiteData | {}): SiteData => {
  let textColor = '#FFFFFF';
  let background = '#1276CE';
  let isLight = false;
  if (!isEmpty(initialSiteData) && 'textColor' in initialSiteData) {
    textColor = initialSiteData.textColor;
    background = initialSiteData.background;
    isLight = JSON.parse((initialSiteData.isLight as unknown) as string);
  }

  const initialCombos = ColorCombos([textColor, background]) as ColorCombosTypes[];
  return {
    background: background,
    textColor: textColor,
    isLight,
    colorCombos: initialCombos,
  };
};

const checkBackgroundLightness = (hex: string): boolean => {
  let light;
  try {
    light = Color(hex).isLight();
  } catch (e) {
    light = true;
  }
  return light;
};

const isValidColor = (value: string): Color | false => {
  let color: Color | false = false;
  try {
    color = Color(value);
  } catch (error) {
    console.error('ColorInput invalid color');
  }
  return color;
};

const createFakeCombination = (color: number[], hex: string): ColorCombinationTypes => {
  return {
    model: 'rgb',
    color: color,
    valpha: 1,
    hex: hex,
    contrast: 1,
    accessibility: { aa: false, aaLarge: false, aaa: false, aaaLarge: false },
  };
};

const createDuplicateCombination = (combos: ColorCombosTypes[]): ColorCombosTypes[] => {
  const color = combos[0].color != null ? combos[0].color : [];
  const dupeCombo = {
    ...combos[0],
    combinations: [createFakeCombination(color, combos[0].hex)],
  };
  return [dupeCombo, dupeCombo];
};

const HomeContext = React.createContext<HomeContextInterface | null>(null);

const useSiteData = (): HomeContextInterface => {
  const context = React.useContext(HomeContext);
  if (!context) {
    throw new Error('useSiteData must be used with SiteDataProvider');
  }
  return context;
};

const SiteDataProvider: React.FunctionComponent<SiteDataProviderProps> = (
  props: SiteDataProviderProps
): React.ReactElement => {
  const [siteData, setSiteData] = React.useState<SiteData>(
    setInitialContext(props.initialSiteData)
  );

  const [isInitial, setIsInitial] = React.useState<boolean>(false);

  const [state] = React.useMemo(
    (): [SiteData, React.Dispatch<SiteData>] => [siteData, setSiteData],
    [siteData]
  );

  const updateHash = debounce((): void => {
    const query = '?' + qs.stringify(state as {});
    window.history.pushState(state, 'Are My Colors Accessible', query);
  }, 200);

  React.useEffect((): void => {
    if (isInitial) {
      updateHash();
    } else {
      setIsInitial(true);
    }
  }, [state]);

  const setNewColorCombo = (textColor: string, backgroundColor: string): void => {
    let newCombos: ColorCombosTypes[] | false = ColorCombos([textColor, backgroundColor]);
    if (newCombos) {
      if (textColor === backgroundColor) {
        newCombos = createDuplicateCombination(newCombos);
      }
      setSiteData({
        ...state,
        background: backgroundColor,
        colorCombos: newCombos,
        isLight: checkBackgroundLightness(backgroundColor),
        textColor: textColor,
      });
    }
  };

  const handleBackgroundColorInputChange = (value: string): void => {
    setSiteData({
      ...state,
      background: value,
    });
    if (isValidColor(value)) {
      setNewColorCombo(state.textColor, value);
    }
  };

  const handleTextColorInputChange = (value: string): void => {
    setSiteData({
      ...state,
      textColor: value,
    });
    if (isValidColor(value)) {
      setNewColorCombo(value, state.background);
    }
  };

  return (
    <HomeContext.Provider
      value={{
        siteData: state,
        handleBackgroundColorInputChange: handleBackgroundColorInputChange,
        handleTextColorInputChange: handleTextColorInputChange,
      }}
      {...props}
    />
  );
};

export { HomeContext, useSiteData, SiteDataProvider };
