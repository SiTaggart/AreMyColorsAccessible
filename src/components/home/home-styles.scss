.home {
    height: 85%;
    position: relative;

    .form {
        bottom: 0;
        left: 0;
        position: absolute;
        right: 0;
        width: 100%;
    }
}
