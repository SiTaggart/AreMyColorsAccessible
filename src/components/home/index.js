import React, { Component, PropTypes } from 'react';
import Colorable from 'colorable';
import './home-styles.scss';
import { Container, LayoutSmall, LayoutLarge } from '../layouts';
import Results from '../results';
import ColorInputs from '../colorInputs';

class Home extends Component {

    constructor(props) {
        super(props);
        this.setBackgroundColor = this.setBackgroundColor.bind(this);
        this.setForegroundColor = this.setForegroundColor.bind(this);
    }

    getColorInfo(background, foreground) {
        let colorInfo;
        try {
            colorInfo = Colorable([foreground, background])[0].combinations[0];
        } catch(e) {
            colorInfo = {
                contrast: 0,
                accessibility: {
                    aaa: false,
                    aa: false,
                    aaaLarge: false,
                    aaLarge: false
                }
            };
        }
        return colorInfo;
    }

    setBackgroundColor(hex) {
        this.props.setBackgroundColor(hex);
    }

    setForegroundColor(hex) {
        this.props.setForegroundColor(hex);
    }

    render() {
        const foreground = this.props.siteData.foreground;
        const background = this.props.siteData.background;
        const colorInfo = this.getColorInfo(background, foreground);

        return (
            <Container className="home">
                <LayoutSmall>
                    <Results {...colorInfo} />
                </LayoutSmall>
                <LayoutLarge>
                    <ColorInputs
                        background={this.props.siteData.background}
                        foreground={this.props.siteData.foreground}
                        isLight={this.props.siteData.isLight}
                        setForegroundColor={this.setForegroundColor}
                        setBackgroundColor={this.setBackgroundColor}
                    />
                </LayoutLarge>
                {this.props.children}
            </Container>
        );
    }
}

Home.propTypes = {
    children: PropTypes.node,
    setBackgroundColor: PropTypes.func,
    setForegroundColor: PropTypes.func,
    siteData: PropTypes.shape({
        background: PropTypes.string,
        foreground: PropTypes.string,
        isLight: PropTypes.bool
    })
};

export default Home;
