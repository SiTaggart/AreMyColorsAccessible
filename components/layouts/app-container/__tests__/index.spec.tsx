/* eslint-env jest */

import React, { ReactElement } from 'react';
import ReactDOM from 'react-dom';
import { shallow, ShallowWrapper } from 'enzyme';
import AppContainer from '..';
import { HomeProps } from '../../../home'; // eslint-disable-line no-unused-vars

describe('AppContainer', (): void => {
  let ChildComponent: React.FunctionComponent<HomeProps> = (): ReactElement<HTMLDivElement> => {
    return <div />;
  };
  let wrapper: ShallowWrapper;
  let instance: AppContainer;

  beforeAll(
    (): void => {
      ChildComponent = function ChildComponent(): JSX.Element {
        return <div />;
      };
    }
  );

  beforeEach(
    (): void => {
      wrapper = shallow(
        <AppContainer title="are my colors accessible">
          {(props): ReactElement => <ChildComponent {...props} />}
        </AppContainer>
      );

      instance = wrapper.instance() as AppContainer;
    }
  );

  it('renders without crashing', (): void => {
    ReactDOM.render(
      <AppContainer title="are my colors accessible">
        {(props): ReactElement => <ChildComponent {...props} />}
      </AppContainer>,
      document.createElement('div')
    );
  });

  it('should clone children with props', (): void => {
    expect(wrapper.find('ChildComponent').getElement().props.siteData).toEqual({
      background: '#1276CE',
      colorCombos: [
        {
          color: [255, 255, 255],
          combinations: [
            {
              accessibility: { aa: true, aaLarge: true, aaa: false, aaaLarge: true },
              color: [18, 118, 206],
              contrast: 4.658034537943552,
              hex: '#1276CE',
              model: 'rgb',
              valpha: 1
            }
          ],
          hex: '#FFFFFF',
          model: 'rgb',
          valpha: 1
        },
        {
          color: [18, 118, 206],
          combinations: [
            {
              accessibility: { aa: true, aaLarge: true, aaa: false, aaaLarge: true },
              color: [255, 255, 255],
              contrast: 4.658034537943552,
              hex: '#FFFFFF',
              model: 'rgb',
              valpha: 1
            }
          ],
          hex: '#1276CE',
          model: 'rgb',
          valpha: 1
        }
      ],
      isLight: false,
      textColor: '#FFFFFF'
    });
    expect(
      typeof wrapper.find('ChildComponent').getElement().props.handleBackgroundColorInputChange
    ).toBe('function');
    expect(
      typeof wrapper.find('ChildComponent').getElement().props.handleBackgroundColorSliderChange
    ).toBe('function');
    expect(
      typeof wrapper.find('ChildComponent').getElement().props.handleTextColorInputChange
    ).toBe('function');
    expect(
      typeof wrapper.find('ChildComponent').getElement().props.handleTextColorSliderChange
    ).toBe('function');
  });

  it('should set state with queryParams', (): void => {
    window.history.pushState(
      {},
      'Test Title',
      '/?background=%235E38A8&isLight=false&textColor=%23FFFFFF'
    );
    const newwrapper = shallow(
      <AppContainer title="are my colors accessible">
        {(props): ReactElement => <ChildComponent {...props} />}
      </AppContainer>
    );
    expect(newwrapper.instance().state).toEqual({
      siteData: {
        background: '#5E38A8',
        colorCombos: [
          {
            color: [255, 255, 255],
            combinations: [
              {
                accessibility: { aa: true, aaLarge: true, aaa: true, aaaLarge: true },
                color: [94, 56, 168],
                contrast: 8.055122537045744,
                hex: '#5E38A8',
                model: 'rgb',
                valpha: 1
              }
            ],
            hex: '#FFFFFF',
            model: 'rgb',
            valpha: 1
          },
          {
            color: [94, 56, 168],
            combinations: [
              {
                accessibility: { aa: true, aaLarge: true, aaa: true, aaaLarge: true },
                color: [255, 255, 255],
                contrast: 8.055122537045744,
                hex: '#FFFFFF',
                model: 'rgb',
                valpha: 1
              }
            ],
            hex: '#5E38A8',
            model: 'rgb',
            valpha: 1
          }
        ],
        isLight: false,
        textColor: '#FFFFFF'
      }
    });
  });

  it('should update state when handleBackgroundColorInputChange is called with a hex', (): void => {
    instance.handleBackgroundColorInputChange('#000');
    expect(instance.state.siteData.background).toEqual('#000');
    expect(instance.state.siteData.colorCombos[1].hex).toEqual('#000000');
  });

  it('should not update colorCombos on state when handleBackgroundColorInputChange is called with a non-valid color', (): void => {
    instance.handleBackgroundColorInputChange('tase848ncdnje9k');
    expect(instance.state.siteData.background).toEqual('tase848ncdnje9k');
    expect(instance.state.siteData.colorCombos[1].hex).toEqual('#5E38A8');
  });

  it('should update state with isLight when handleBackgroundColorInputChange is called with a light color', (): void => {
    instance.handleBackgroundColorInputChange('#ccc');
    expect(instance.state.siteData.isLight).toEqual(true);
  });

  it('should update state when handleTextColorInputChange is called with a hex', (): void => {
    instance.handleTextColorInputChange('#fefefe');
    expect(instance.state.siteData.textColor).toEqual('#fefefe');
    expect(instance.state.siteData.colorCombos[0].hex).toEqual('#FEFEFE');
  });

  it('should update the url with the state when updateHash is called', (): void => {
    instance.handleBackgroundColorInputChange('#777');
    instance.handleTextColorInputChange('#ddd');
    instance.updateHash();
    expect(window.location.search).toEqual(
      '?background=%23777&colorCombos=%5Bobject%20Object%5D&colorCombos=%5Bobject%20Object%5D&isLight=false&textColor=%23ddd'
    );
  });
});
