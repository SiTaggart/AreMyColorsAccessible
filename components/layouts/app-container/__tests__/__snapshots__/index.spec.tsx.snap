// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContainer changes the footer link colors based on isLight prop 1`] = `
@media (min-width:768px) {
  .emotion-18 {
    height: 100%;
  }
}

.emotion-16 {
  padding: 5rem 0 1rem;
  text-align: center;
  width: 100%;
}

.emotion-12 {
  list-style: none;
  margin: 0 0 1rem;
  padding: 0;
}

.emotion-2 {
  display: inline-block;
  padding: 0 1rem;
}

.emotion-0 {
  color: currentColor;
  -webkit-transition: color 400ms ease-in;
  transition: color 400ms ease-in;
}

<div
  className="emotion-18 emotion-19"
>
  <div />
  <footer
    className="emotion-16 emotion-17"
  >
    <nav
      className="emotion-14 emotion-15"
    >
      <ul
        className="emotion-12 emotion-13"
      >
        <li
          className="emotion-2 emotion-3"
        >
          <a
            className="emotion-0 emotion-1"
            href="/"
            onClick={[Function]}
            style={
              Object {
                "color": "#343334",
              }
            }
          >
            Home
          </a>
        </li>
        <li
          className="emotion-2 emotion-3"
        >
          <a
            className="emotion-0 emotion-1"
            href="/palette"
            onClick={[Function]}
            style={
              Object {
                "color": "#343334",
              }
            }
          >
            Palette
          </a>
        </li>
        <li
          className="emotion-2 emotion-3"
        >
          <a
            className="emotion-0 emotion-1"
            href="/about"
            onClick={[Function]}
            style={
              Object {
                "color": "#343334",
              }
            }
          >
            About
          </a>
        </li>
      </ul>
    </nav>
  </footer>
</div>
`;

exports[`AppContainer renders without crashing 1`] = `
@media (min-width:768px) {
  .emotion-18 {
    height: 100%;
  }
}

.emotion-16 {
  padding: 5rem 0 1rem;
  text-align: center;
  width: 100%;
}

.emotion-12 {
  list-style: none;
  margin: 0 0 1rem;
  padding: 0;
}

.emotion-2 {
  display: inline-block;
  padding: 0 1rem;
}

.emotion-0 {
  color: currentColor;
  -webkit-transition: color 400ms ease-in;
  transition: color 400ms ease-in;
}

<div
  className="emotion-18 emotion-19"
>
  <div />
  <footer
    className="emotion-16 emotion-17"
  >
    <nav
      className="emotion-14 emotion-15"
    >
      <ul
        className="emotion-12 emotion-13"
      >
        <li
          className="emotion-2 emotion-3"
        >
          <a
            className="emotion-0 emotion-1"
            href="/"
            onClick={[Function]}
            style={
              Object {
                "color": "#fff",
              }
            }
          >
            Home
          </a>
        </li>
        <li
          className="emotion-2 emotion-3"
        >
          <a
            className="emotion-0 emotion-1"
            href="/palette"
            onClick={[Function]}
            style={
              Object {
                "color": "#fff",
              }
            }
          >
            Palette
          </a>
        </li>
        <li
          className="emotion-2 emotion-3"
        >
          <a
            className="emotion-0 emotion-1"
            href="/about"
            onClick={[Function]}
            style={
              Object {
                "color": "#fff",
              }
            }
          >
            About
          </a>
        </li>
      </ul>
    </nav>
  </footer>
</div>
`;
